### ISSUE TITLE
Generalize proxy to allow connections to white-listed URLs

### ISSUE THREAD

**Issue by rbrecheisen** [IKNL/vantage6-node#34]
Based on our discussion on Discord. So, to reiterate we'd like to allow the algorithm container to connect to XNAT directly instead of having to go through an ETL that pulls the data and stores it inside the Docker volume.

Cheers, Ralph

**Comment by frankcorneliusmartin** [comment]
To allow the algorithm containers to reach external services we could extend the local proxy (https://github.com/IKNL/vantage6-node/blob/master/vantage6/node/proxy_server.py) to also allow to reach other endpoints than just the central server. 

We could add an extra option in the node configuration files:
```yaml
whitelist:
    - 10.0.0.1
    - ....
```

**Comment by frankcorneliusmartin** [comment]
After a chat with Pedro Mateus, the solution above does not profide a solution for external non-api based services. As the `local proxy` is basically a API forwarder. 

A more general solution would be to forward any kind of protocol. An 💡 would be to attach host-names to a proxy docker container that forwards these requests without alteration. The attached host-names would be the hosts from the whitelist. So the only requests that reach this proxy container are accepted resources. 

_ Related: this could also be used for the node-node communication update in version 3. _

**Comment by rbrecheisen** [comment]
Hi Frank, I wanted to leave you a message on Discord but I can't seem to login anymore. It's showing a page asking me to claim my account but all the fields are disabled...

As to the above, I'm not sure I understand what you're saying. As a vantage6 user I'd like to connect to XNAT from the algorithm container but your suggestion above does not tell me much :)

**Comment by frankcorneliusmartin** [comment]
I don't know what to do about the Discord issue, you might want to reach out to the Discord support? 😕

Ok I'll try to be more complete. So my initial suggestion was to extend the local proxy to accept a whitelist. (https://github.com/IKNL/vantage6-node/blob/master/vantage6/node/proxy_server.py). If I understood correctly, the XNAT service has an API. So this initial would solve your use case. 

However, there are services that do not have a rest API. For example a postgress-database does not have one by default (it communicates using a TCP socket) and I dont think it is desirable to do so. So in order to support other use-cases  then the one you are proposing we need to generalize the communication protocol.  

*I am not sure if this is possible*, but my guts says it is. But the idea is to create a container having the hostnames (=aliases) from the user configurable whitelist. Then the container can make requests to this domain or ip, and thereby reaching the Proxy Container. This proxy container is reached -and only reached- when the destination is on the whitelist. In that case the proxy containers forwards the request to `the internet`. It feels to me like some sort of `Reverse proxy` but i am not an expert in this field.

So to go forward, we need to investigate how such a container would look like and if this approach has any drawbacks.  

Does that make more sense?

**Comment by rbrecheisen** [comment]
Hi Frank,

Thanks for the extra explanation! I think I more or less understand the
issue :)
I will wait for any progress in this direction.

Cheers,

Ralph

**Comment by frankcorneliusmartin** [comment]
@rbrecheisen I am not sure if you are aware of the Fork from pedro: https://github.com/pedro-cmat/vantage6-node
I think his version is able to reach API's outside of the Docker network. He is also on Discord, so you can reach him there if you have any questions.

**Comment by rbrecheisen** [comment]
Right! I know Pedro already. He's also in Maastricht and I work together
with him on getting vantage6 working with XNAT

Ralph

**Comment by frankcorneliusmartin** [comment]
It seems that NGINX is able to help us out. 

> we could also use ip tables for this purpose, but I intend to use NGINX also as a reverse proxy for the peer-to-peer connections between data stations in v3.0.0

Building Docker image should give us a forward proxy from the algorithm containers. I have not tested this within the infrastructure. But I guess Ill do that in the next few days.

**Comment by frankcorneliusmartin** [comment]
Started working on this in https://github.com/iknl/vantage6-node/tree/feature/eduvpn-connect

This can be implemented using `iptables` that are already used by the eduvpn setup. See the CARRIER project: https://github.com/CARRIER-project/algorithm-container-network

**Comment by bartvanb** [comment]
There are some security concerns with whitelisting in the current setup and will therefore postpone implementation of this feature.
